lib/components/user/notification-prefs-pane.js

```diff
@@ -28,85 +36,393 @@ const allowedNotificationChannels = [
+                  // in order for Formik to correcly process the changes.
+                  onBlur={handleBlur}
+                  onChange={handleChange}
+                  value={type}
+                >
+                  {text}
+                </ToggleButton>
+              ))}
+            </ToggleButtonGroup>
+          </ButtonToolbar>
+        </FormGroup>
+        <Details>
+          {notificationChannel === 'email' && (
+            <FormGroup>
+              <ControlLabel>Notification emails will be sent to:</ControlLabel>
+              <FormControl.Static>{email}</FormControl.Static>
+            </FormGroup>
+          )}
+          {notificationChannel === 'sms' && (
+            <Formik
+              // Phone number needs to be validated on change.
+              validateOnChange
+              validationSchema={codeValidationSchema}
+              initialValues={initialFormikValues}
+            >
+              {
+                // Pass Formik props to the component rendered so Formik can manage its validation.
+                // (The validation for this component is independent of the validation set in UserAccountScreen.)
+                innerProps => {
+                  return (
+                    <PhoneNumberEditor
+                      {...innerProps}
+                      initialPhoneNumber={phoneNumber}
+                      initialPhoneNumberVerified={isPhoneNumberVerified}
+                      onRequestCode={onRequestPhoneVerificationCode}
+                      onSubmitCode={onSendPhoneVerificationCode}
+                      onVerified={this._handleOnPhoneVerified}
+                      phoneFormatOptions={phoneFormatOptions}
+                    />
+                  )
+                }
+              }
+            </Formik>
+          )}
+        </Details>
+      </div>
+    )
+  }
+}
+
+export default NotificationPrefsPane
+
+/**
+ * Sub-component that handles phone number and validation code editing and validation intricacies.
+ */
+class PhoneNumberEditor extends Component {
+  constructor (props) {
+    super(props)
+
+    const { initialPhoneNumber, initialPhoneNumberVerified } = props
+    this.state = {
+      // Holds the last known verified phone number.
+      currentNumber: initialPhoneNumber,
+
+      // The verified status of the above number.
+      currentNumberVerified: initialPhoneNumberVerified,
+
+      // If true, phone number is being edited.
+      // For new users, render component in editing state.
+      isEditing: isBlank(initialPhoneNumber),
+
+      // If true, a phone verification request has been sent and the UI is awaiting the user sending the code.
+      isPending: false,
+
+      // Holds the new phone number (+15555550123 format) entered by the user
+      // (outside of Formik because (Phone)Input does not have a standard onChange event or simple valitity test).
+      newPhoneNumber: ''
+    }
```

Comment:
Pretty much all of these component state variables are changed when application state changes. Therefore, they should probably be application state variables instead.

Reply:
all except `isEditing` were removed by https://github.com/opentripplanner/otp-react-redux/pull/224/commits/bc43d29931e3907bca16067f26d718075ad0e6e3.